### PR TITLE
Add useStyletron hook and default no-op engine

### PR DESCRIPTION
**Pull request by tajo**
solves https://github.com/styletron/styletron/issues/256, https://github.com/styletron/styletron/issues/284#issuecomment-467194731

**EDIT: Descoped, explained [bellow](https://github.com/styletron/styletron/pull/304#issuecomment-479652251).**

# useStyletron

`useStyletron` is a React [hook](https://reactjs.org/docs/hooks-intro.html). 

This API introduces a very direct way of styling components (it's more discussed in https://github.com/styletron/styletron/issues/284#issuecomment-467194731). It should be handy for non-customizable components since it's pretty concise (no need to name the styled components) and prevents multiple pitfalls (ref forwarding, prop filtering...).

**It also supports theming**.

```jsx
import { Provider, useStyletron } from "styletron-react";

export default () => {
  const [ css ] = useStyletron();
  return (
    <div className={css({ color: 'blue' })}>Blue text</div>
  );
};
```

# No-op engine

If you forget to wrap your application with `<Provider>`, Styletron switches to a no-op mode and prints out a warning (instead of crashing). This is nice when writing tests. You don't need to setup Styletron if you don't care about styles. However, you should set `NODE_ENV = test` to prevent the warnings.

# Supporting only React 16.3+ / 16.8+

**This PR should be landed as a part of v5 milestone.** Why?

I believe it makes sense to drop the support for React v15 since v16 was released [18 months ago](https://github.com/facebook/react/releases/tag/v16.0.0). 

v16.3 introduces ref forwarding and createContext APIs. This PR removes `create-react-context` polyfill to save some bytes and we are going to implement ref forwarding in a few weeks as well. So 16.3+ is hard minimum.

If you want to use the `useStyletron` hook, you need to be on v16.8. Since it should be fairly easy to upgrade React from v16.3 to v16.8, I set the peerDependency to v16.8.

# Todos

- debugEngine is broken in the `useStyletron` API, I need to look more into how the whole thing works (I was not able to make the sourcemaps work yet even for the current version). Also, we could fix it later. 
- since this will be released as v5, docs will be updated later

### CODE REVIEW

**Comment by rtsao** [comment]
# Theming

Unfortunately, I don't think type inference for themes will be possible in APIs where a `React.createContext` value isn't imported -- the very thing we are trying to avoid in the first place.

```js
const ThemeContext = React.createContext(someTheme);

// $theme type can be inferred
const StyledComponent = styled("div", ({ $theme }) => ({
  color: $theme.textColor
}), ThemeContext);

// theme type can be inferred
function HookComponent() {
  const { theme } useContext(ThemeContext);
  const { css } = useStyletron();
}
```

For consistency with `useStyletron`, potentially the signature of `styled` could be changed to look like:

```js
const StyledComponent = styled("div", (props, [theme, other]) => ({
  color: theme.textColor
}), [ThemeContext, OtherContext]);
```

Of course, the tradeoff is the extra burden of importing these context values around. This is certainly an idiomatic React pattern, but is potentially not worth the burden. That said, I think there is a real benefit (aside from better typing) to moving theming out of core in the sense that there can be multiple themes.

Imagine a web application using two component libraries implemented with Styletron. If theming is in core and both libraries use themes, it is essentially impossible for both of these libraries to coexist without breaking each other or having to know about each other. Of course, multiple themes could be accommodated if different themes were always namespaced in the core theme context, but this is merely convention and has the same drawbacks as legacy React context (naming collisions, etc.)

So realistically, as a convention, theme will probably always need to be namespaced. For example:

```js
const Box = styled("div", ({ $theme }) => ({
  color: $theme.myTheme.textColor
}));
```

So I think this not exactly ideal either. Also, component portability and isolation has always been a paramount concern for Styletron, which is definitely aligned with the approach of theming via vanilla React context.

In summary this is how I see this tradeoff:

## Theming in core

#### Benefits
- Don't have to import React context values when consuming context. Since these are often relative paths, this can be burdensome

#### Drawbacks
- Theme is untyped by default. Type annotations for theme must be manually added, and potentially involves importing a type from elsewhere
- Potential theme collisions. Relies on namespacing conventions to work properly.
- No way to consume arbitrary non-theme context in styles when using `styled`

## React Context-based theming

#### Benefits
- Themes (and other context usage) are automatically inferred with no need for annotations
- No possibility of theme collisions. Components can be completely unaware of different themes.
- Theme is no different than any other context. Straightforward to consume arbitrary non-theme context in styles when using `styled`
- Nothing new to learn with `useStyletron` hook. Themes are just regular context.

#### Drawbacks
- Must import React context values when consuming context. Since these are often relative paths, this can be burdensome

**Comment by chasestarr** [comment]
Do we consider the noop-engine to be a breaking change? It would cause churn right, because many devs rely on the shape in snapshot tests? Not sure of the semver stance here. Those changes generally lgtm, but would be interested to see an isolated pr for it if possible.

I'm on-board with the peer-dep bump. 👍

Can `useStyletron` land without the theme integration? An approach like below should still be possible without theme. Would adding theme support require a major release? I think everyone can agree on the other features included in the pr. If it's easy to add later, it may be worth pulling out for now.

```js
import React from 'react';
import { useStyletron } from "styletron-react";
import MyThemeContext from '../my-theme.js';

export default () => {
  const { css } = useStyletron();
  const { theme } = useContext(MyThemeContext);
  return (
      <div className={css({ color: theme.textColor })}>Blue text</div>
  );
};
```

**Comment by rtsao** [comment]
The React peer dependency change (required for hooks support) is definitely a breaking change. I think adding a default Noop engine is probably technically not breaking, but I think releasing this with some other breaking change makes sense.

I think adding theme support later (depending on the API) can most likely be a minor change.

**Comment by tajo** [comment]
## Unstructured random thoughts

**Great points and the summary of benefits / drawbacks! There is nothing I could really argue with.**

Frankly, this PR is about some minor quality of life improvements. It doesn't fixes bugs or changes things fundamentally. So I think we can't justify breaking some current APIs / patterns just to land this. It would cancel all the goodwill.

> Must import React context values when consuming context. Since these are often relative paths, this can be burdensome

I still see this as a big hurdle. It's already hard to make developers using the `$theme` values exclusively since it's easier to just hardcode `8px` and the `$theme` variable is right there (I am guilty of this too). The extra relative import would add a lot of friction. 

My initial goal was to deprecate `baseui/styled` with this PR since it could be just replaced by Styletron's `styled`. This is already well established pattern

```jsx
const Box = styled("div", ({ $theme }) => ({
  color: $theme.textColor
}));
```

even outside of `baseui` components, so I think at this point it's not realistic to yank the theme out of it and add `useContext` everywhere. That would make devs upset.

Also, the same pattern (`ThemeProvider` / `props.theme`) is used by all other major CSS in JS libs as well, so many developers are already familiar with it. 

## Types

Yea, it seems there is no way how you could type the theme out of the box and indeed, the other libs just go with `any`. 

However, if we remove the assumption that `baseui/styled` should cease to exist, I think we can preserve it as a light wrapper that just adds the types.  So

- by default, the `theme` is typed as `any` and that might be fine if you don't use Flow/TS or just don't care
- if you want to get it typed, you need to re-export `styled` and `useStyletron` so you can use your theme to set the correct type

This also addresses some drawbacks:

> Imagine a web application using two component libraries implemented with Styletron. 

you can namespace your theme

```jsx
<Provider theme={{ baseui: theme }}>
```

and remove the `baseui` part when using `baseui/styled`. Although, this sounds more like a theoretical problem.

## TLDR;
- I really don't want to change the current APIs / patterns. Especially if it doesn't simplify the usage. `theme = useContext()` just adds more friction. Although, I can't argue that it's a cleaner approach.
- If the `styled` function has `props.$theme`, `useStyletron` should provide `theme` as well. They should behave similarly.
- The theme can be typed as `any`, you can reexport `styled` and `useStyletron` to type it properly.
- **I am down to land this PR without the theme support** and keeping this functionality only in `baseui/styled` and `baseui/useStyletron`. Although, I still think it would be valuable to all Styletron users.

**Comment by rtsao** [comment]
### Interface changes

> I really don't want to change the current APIs / patterns.

I would definitely have concerns about usage of `$theme` without namespacing. Do you agree that namespacing the baseUI theme would be a practical necessity as part of a potential migration from `baseui/styled` to vanilla `styled` (in order to avoid collisions)? For example:

```diff
- import {styled} from 'baseui';
+ import {styled} from 'styletron-react';

  const BlueDiv = styled('div', ({$theme}) => ({
-  color: $theme.colors.primary,
+  color: $theme.baseui.colors.primary,
  }));
```

If so, then at the very least, there will have to be a codemod that does the above transformation. 

However, potentially the codemod could instead do:

```diff
- import {styled} from 'baseui';
+ import {styled, ThemeContext} from 'baseui';

- const BlueDiv = styled('div', ({$theme}) => ({
+ const BlueDiv = styled('div', (props, $theme) => ({
    color: $theme.colors.primary,
- }));
+ }), ThemeContext);
```

Standalone, this looks like:
```js
import {styled, ThemeContext} from 'baseui';

const BlueDiv = styled('div', (props, $theme) => ({
  color: $theme.colors.primary,
}), ThemeContext);
```

There's *slightly* more code, but I personally don't think it constitutes a meaningful penalty in terms of boilerplate, especially given the other advantages of an API like this. And assuming there's a codemod, there's very little pain for developers.

Furthermore, there'd be no need for an immediate breakage of the existing interface. It could be deprecated and then removed in the next major version. For example:

```js
import {styled} from "baseui";

// Works, but triggers deprecation warning. Tells developers how to run codemod
const Box = styled("div", ({ $theme }) => ({
  color: $theme.textColor
}));
```

### `useContext` friction

> theme = useContext() just adds more friction

I think this friction could potentially be slightly reduced if `useStyletron` also (optionally) serves as a vanilla context consumer. For example:

```js
import {useStyletron, ThemeContext} from "baseui";

function Component(props) {
  const [css, theme] = useStyletron(ThemeContext);
  /* ... */
}
```

This particular `useStyletron` hook interface would be a potential reason to hold off on releasing it until we sort out theming, since this would be a breaking change from `const {css} = useStyletron();`.

**Comment by tajo** [comment]
All your examples import `from 'baseui'`, is that a typo? Like here:

```jsx
import {styled, ThemeContext} from 'baseui';

const BlueDiv = styled('div', (props, $theme) => ({
  color: $theme.colors.primary,
}), ThemeContext);
```

Otherwise, I am not sure why to expose the `ThemeContext` "binding" if the user is already using `baseui/styled`. We can just do (keep doing) this for them.

>  Do you agree that namespacing the baseUI theme would be a practical necessity as part of a potential migration from baseui/styled to vanilla styled (in order to avoid collisions)?

Yea. That's perfectly fine to do.

**So basically are there two ways forward:**

1. `baseui/styled` stays in place with the same API, `baseui/useStyletron` is introduced and "magically" provides `theme` as well
2. `baseui/styled` is deprecated, everyone should import `styled` and `useStyletron` from `styletron-react` only and pass the `ThemeContext` manually

If we go with `1.` we would still implement this for `styletron-react`:

```js
const [css, theme] = useStyletron(ThemeContext);
const BlueDiv = styled('div', (props, $theme) => ({
  color: $theme.colors.primary,
}), ThemeContext);
```

I am now more inclining to keep offering `baseui` versions of `styled` / `useStyletron` as a simpler alternatives / shortcuts. 

One use-case to thing about: **What if I want to dynamically switch between two themes** (dark mode vs light mode)? Ideally I should be able to make that switch in a single place - the Provider? How would this accommodate that?

```js
import {styled} from 'styletron-react';
import {ThemeContext} from 'baseui';

const BlueDiv = styled('div', (props, $theme) => ({
  color: $theme.colors.primary,
}), ThemeContext);
```

This is also probably the only good reason why not to just import the theme object directly:

```js
import {styled, theme} from 'baseui';

const BlueDiv = styled('div', (props => ({
  color: theme.colors.primary,
}));
```

**Comment by tajo** [comment]
Discussing this with @rtsao 1-1 and here's the conclusion (correct me if wrong!):

We'll do the the most conservative thing and keep the theming out of Styletron for now. This PR will just add this API:

```jsx
const [css] = useStyletron();
```

The array is used so we can potentially return more parameters without a breaking change.

It turns out it makes a lot of sense keeping the theming component library specific so **you can get precise types** and avoid any namespace collisions. Also, it's quite simple these days to use React Context APIs directly (especially with the `useContext` hook), so there's not much benefit to abstracting it by Styletron. Especially not when you consider drawbacks (losing types, namespacing...).

For those who follow [BaseUI](https://github.com/uber-web/baseui), we will keep providing the `styled` function as is and `useStyletron` will be introduced so you can do:

```jsx
import { useStyletron } from 'baseui';

const [css, theme] = useStyletron();
```

File: packages/styletron-react/src/index.js [hunk 1]
```diff
@@ -107,6 +114,52 @@ type createStyledOpts = {
...
+  }
+}
+
+export function useStyletron() {
+  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
+  const debugEngine = React.useContext(DebugEngineContext);
+  const hydrating = React.useContext(HydrationContext);
+  return [
+    function css(style: StyleObject) {
+      checkNoopEngine(styletronEngine);
```
Comment (rtsao) [inline packages/styletron-react/src/index.js]: Do you think it is necessary for this check every `css` call?  Or should this be hoisted up a level to when `useStyletron` is called.

File: packages/styletron-react/src/index.js [hunk 2]
```diff
@@ -107,6 +114,52 @@ type createStyledOpts = {
...
+      const className = driver(style, styletronEngine);
+      if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {
+        const {stack, message} = new Error("stacktrace source");
+        const debugClassName = debugEngine.debug({
```
Comment (rtsao) [inline packages/styletron-react/src/index.js]: I'm wondering if this could be put into `useMemo` so this is only called once (rather than for each `css` call. This might save some unnecessary work.

File: packages/styletron-react/src/index.js (hunk 1 above)
Comment (tajo) [reply packages/styletron-react/src/index.js]: good point, moved

File: packages/styletron-react/src/index.js (hunk 2 above)
Comment (tajo) [reply packages/styletron-react/src/index.js]: I played with it a bit but no luck. The problem is that it get memoized when `hydrating` or if I remove that from the guarding condition `if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {}`, then I get the mismatch server/client error. HMR just works (with or w/o the memo). It even fixes the last sourceMap link problem and starts pointing to the right place :)

Screen Shot 2019-04-05 at 3 46 34 PM

Anyway, here's what I ended up with before I gave up:

```jsx
export function useStyletron() {
  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
  const debugEngine = React.useContext(DebugEngineContext);
  const hydrating = React.useContext(HydrationContext);
  checkNoopEngine(styletronEngine);
  return [
    function css(style: StyleObject) {
      const className = driver(style, styletronEngine);
      const debugClassName =
        React.useMemo(
          () => {
            if (__BROWSER__ && __DEV__ && debugEngine) {
              const {stack, message} = new Error("stacktrace source");
              return debugEngine.debug({
                stackInfo: {stack, message},
                stackIndex: 5,
              });
            }
            return "";
          },
          [className],
        ) || "";
      if (debugClassName !== "") {
        return `${debugClassName} ${className}`;
      }
      return className;
    },
  ];
}
```

**It has two problems:**
- It also doesn't add the `debug` className on the initial render (not sure why)
- server/client mismatch warning

Also, not sure if `className` makes completely sense as useMemo's dep in this case. Note that the guarding condition needs to be inside of the memo hook, otherwise React explodes.

Comment (rtsao) [reply packages/styletron-react/src/index.js]: What about something like:

```js
export function useStyletron() {
  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
  const debugEngine = React.useContext(DebugEngineContext);
  const hydrating = React.useContext(HydrationContext);
  checkNoopEngine(styletronEngine);
  return [
    function css(style: StyleObject) {
      const className = driver(style, styletronEngine);
      // Use ref for error to avoid having stack trace inside of hooks
      const errorRef = useRef(new Error("stacktrace source"));
      const debugClassName = useMemo(
        () => {
          if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {
            const { stack, message } = errorRef.current;
            return debugEngine.debug({
              stackInfo: { stack, message },
              stackIndex: 1
            });
          }
          return "";
        },
        [hydrating]
      );
      return debugClassName ? `${debugClassName} ${className}` : className;
    }
  ];
}
```

Comment (tajo) [reply packages/styletron-react/src/index.js]: That works great! Thanks! The PR is updated.

File: packages/styletron-react/src/index.js [hunk 3]
```diff
@@ -30,16 +30,23 @@ import {addDebugMetadata, DebugEngine} from "./dev-tool.js";
...
 export {DebugEngine};
 
-const StyletronContext: Context<any> = createReactContext();
-const HydrationContext = createReactContext(false);
-const DebugEngineContext = createReactContext();
+const StyletronContext = React.createContext<StandardEngine>({
+  renderStyle: () => "",
+  renderKeyframes: () => "",
+  renderFontFace: () => "",
+  noopEngine: true,
```
Comment (rtsao) [inline packages/styletron-react/src/index.js]: I think rather than duck typing via an extra `noopEngine` property, it may be more clear to assign this object to a variable and then check via reference equality in `checkNoopEngine`.

At first I thought `noopEngine` might have something to do with enabling/disabling this feature (but that'd be a bit nonsensical). Also since this is basically a "private" property, it'd be nice to not have it show up in the `StandardEngine` type.

Comment (tajo) [reply packages/styletron-react/src/index.js]: Good point! Updated.
